web/src/watchdog.ts

```diff
@@ -71,4 +83,36 @@ export function on_unsuspend(f: () => void): void {
     unsuspend_callbacks.push(f);
 }
 
-setInterval(check_for_unsuspend, CHECK_FREQUENCY_MILLISECONDS);
+/**
+ * Starts the watchdog event listeners.
+ *
+ * This function should be called exactly once during the application
+ * initialization (see server_events.js).
+ */
+export function initialize(): void {
+    setInterval(check_for_unsuspend, CHECK_FREQUENCY_MILLISECONDS);
+
+    // The Page Lifecycle API "resume" event.
+    // This handles when the page is "resumed" from a frozen state, such as when
+    // the user switches back to a tab that Chrome has discarded/frozen to save memory.
+    //
+    // We check for `document` existence to avoid breaking Node.js-based tests that
+    // import this module but don't implement the full DOM API.
+    if (typeof document !== "undefined" && typeof document.addEventListener === "function") {
+        document.addEventListener("resume", () => {
+            check_for_unsuspend();
+        });
+    }
+
+    // The BFCache "pageshow" event.
+    // This handles when the page is restored from the Back-Forward Cache (BFCache),
+    // e.g. when the user navigates back to this page.
+    // We check for `event.persisted` to ensure we only trigger on a restore, not a fresh load.
+    if (typeof window !== "undefined" && typeof window.addEventListener === "function") {
+        window.addEventListener("pageshow", (event) => {
+            if (event.persisted) {
+                check_for_unsuspend();
+            }
+        });
+    }
+}
```

Comment:
Which of these two listeners did you actually hand-verify?

A thing to verify is temporarily deleting the existing watchdog `setInterval` call, and checking which browsers actually will handle suspend correctly due to the `resume` handler. I'd expect Chrome and not Firefox, for example.